### PR TITLE
Clients Groups

### PR DESCRIPTION
**Pull request by SebaKas21**
Hi @dawiddworak88 

On your hands i give you first stage of our clients groups.

Crud and assign clients to groups work perfectly.
Tiles navigation is now draggable but i know that should have some improvements (i must experiment)

### CODE REVIEW

File: be/src/Project/Services/Client/Client.Api/Infrastructure/Groups/Entities/Group.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using Foundation.GenericRepository.Entities;
+using System.Collections.Generic;
+
+namespace Client.Api.Infrastructure.Groups.Entities
+{
+    public class Group : Entity
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Infrastructure/Groups/Entities/Group.cs]: **suggestion:** Rename to ClientGroup

File: be/src/Project/Services/Client/Client.Api/Infrastructure/ClientContext.cs [hunk 2]
```diff
@@ -16,6 +17,8 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
 
         public DbSet<Clients.Entities.Client> Clients { get; set; }
         public DbSet<Address> Addresses { get; set; }
-
+        public DbSet<Group> Groups { get; set; }
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Infrastructure/ClientContext.cs]: **suggestion:** Rename to public DbSet<ClientGroup> ClientGroups { get; set; }

File: be/src/Project/Services/Client/Client.Api/Infrastructure/ClientContext.cs [hunk 3]
```diff
@@ -16,6 +17,8 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
...
+        public DbSet<GroupTranslation> GroupTranslations { get; set; }
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Infrastructure/ClientContext.cs]: **suggestion:** Rename to DbSet<ClientGroupTranslation> ClientGroupTranslations { get; set; }

File: be/src/Infrastructure/ApiExtensions/Foundation.ApiExtensions/Definitions/ApiConstants.cs [hunk 4]
```diff
@@ -13,6 +13,11 @@ public struct Catalog
             public static readonly string ProductSuggestionsApiEndpoint = "/api/v1/productsuggestions";
         }
 
+        public struct Groups
+        {
+            public static readonly string GroupsApiEndpoint = "/api/v1/groups";
```
Comment (dawiddworak88) [inline be/src/Infrastructure/ApiExtensions/Foundation.ApiExtensions/Definitions/ApiConstants.cs]: **suggestions:** Create a 'Client' struct and put this and other client-related endpoints there

File: be/src/Infrastructure/Localization/Foundation.Localization/Resources/ClientResources.de.resx [hunk 5]
```diff
@@ -144,13 +144,40 @@
...
     <value>Der Kunde wurde erfolgreich gespeichert.</value>
   </data>
+  <data name="ClientsGroups" xml:space="preserve">
+    <value>Kundengruppen</value>
+  </data>
   <data name="EditClient" xml:space="preserve">
     <value>Kunde bearbeiten</value>
   </data>
+  <data name="EditGroup" xml:space="preserve">
+    <value>Gruppe bearbeiten</value>
```
Comment (dawiddworak88) [inline be/src/Infrastructure/Localization/Foundation.Localization/Resources/ClientResources.de.resx]: **suggestion:** Rename Gruppe to Kundengruppe carefully everywhere

File: be/src/Infrastructure/Localization/Foundation.Localization/Resources/ClientResources.en.resx [hunk 6]
```diff
@@ -142,15 +142,42 @@
...
+    <value>Client has been saved successfully.</value>
+  </data>
+  <data name="ClientsGroups" xml:space="preserve">
+    <value>Clients groups</value>
   </data>
   <data name="EditClient" xml:space="preserve">
     <value>Edit client</value>
   </data>
+  <data name="EditGroup" xml:space="preserve">
+    <value>Edit group</value>
```
Comment (dawiddworak88) [inline be/src/Infrastructure/Localization/Foundation.Localization/Resources/ClientResources.en.resx]: **suggestion:** Rename group to client group everywhere

File: be/src/Project/Services/Client/Client.Api/DependencyInjection/CompositionRoot.cs [hunk 7]
```diff
@@ -11,6 +12,7 @@ public static class CompositionRoot
         public static void RegisterClientApiDependencies(this IServiceCollection services)
         {
             services.AddScoped<IClientsService, ClientsService>();
+            services.AddScoped<IGroupsService, GroupsService>();
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/DependencyInjection/CompositionRoot.cs]: **suggestion:** Rename to services.AddScoped<IClientGroupsService, ClientGroupsService>();

File: be/src/Project/Services/Client/Client.Api/Services/Clients/ClientsService.cs [hunk 8]
```diff
@@ -29,45 +30,77 @@ public class ClientsService : IClientsService
...
-                              Email = c.Email,
-                              CommunicationLanguage = c.Language,
-                              PhoneNumber = c.PhoneNumber,
-                              LastModifiedDate = c.LastModifiedDate,
-                              CreatedDate = c.CreatedDate
-                          };
+            var clients = this.context.Clients.Where(x => x.IsActive);
 
-            if (!string.IsNullOrWhiteSpace(model.SearchTerm))
+            if (string.IsNullOrWhiteSpace(model.SearchTerm) is false)
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Clients/ClientsService.cs:35]: **praise:** very nice

File: be/src/Project/Services/Client/Client.Api/Services/Clients/ClientsService.cs [hunk 9]
```diff
@@ -29,45 +30,77 @@ public class ClientsService : IClientsService
...
+                    Id = client.Id,
+                    Name = client.Name,
+                    Email = client.Email,
+                    CommunicationLanguage = client.Language,
+                    PhoneNumber = client.PhoneNumber,
+                    LastModifiedDate = client.LastModifiedDate,
+                    CreatedDate = client.CreatedDate
+                };
+
+                var clientGroups = this.context.ClientsGroups.Where(x => x.ClientId == client.Id && x.IsActive);
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Clients/ClientsService.cs]: **suggestion:** it can be changed to item.Groups = this.context.ClientsGroups.Where(x => x.ClientId == client.Id && x.IsActive).Select(x => x.GroupId)

File: be/src/Project/Services/Client/Client.Api/Services/Clients/ClientsService.cs [hunk 10]
```diff
@@ -29,45 +30,77 @@ public class ClientsService : IClientsService
...
+                Id = existingClient.Id,
+                Name = existingClient.Name,
+                Email = existingClient.Email,
+                CommunicationLanguage = existingClient.Language,
+                PhoneNumber = existingClient.PhoneNumber,
+                LastModifiedDate = existingClient.LastModifiedDate,
+                CreatedDate = existingClient.CreatedDate
+            };
+
+            var clientGroups = this.context.ClientsGroups.Where(x => x.ClientId == existingClient.Id && x.IsActive);
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Clients/ClientsService.cs]: suggestion: the same here

File: be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+            this.context.Groups.Add(group.FillCommonProperties());
+            var groupTranslation = new GroupTranslation
+            {
+                Name = model.Name,
+                Language = model.Language,
+                GroupId = group.Id
+            };
+
+            this.context.GroupTranslations.Add(groupTranslation.FillCommonProperties());
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs]: **suggestion (non-blocking):** redundant new line

File: be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        public async Task DeleteAsync(DeleteGroupServiceModel model)
+        {
+            var group = await this.context.Groups.FirstOrDefaultAsync(x => x.Id == model.Id && x.IsActive);
+
+            if (group is null)
+            {
+                throw new CustomException(this.clientLocalizer.GetString("GroupNotFound"), (int)HttpStatusCode.NotFound);
+            }
+
+            if (await this.context.ClientsGroups.AnyAsync(x => x.GroupId == model.Id && x.IsActive))
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs]: **praise:** bravo

File: be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Client.Api.Services.Groups
+{
+    public class GroupsService : IGroupsService
+    {
+        private readonly ClientContext context;
+        private readonly IStringLocalizer<ClientResources> clientLocalizer;
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs]: **suggestion:** this is most probably not needed here because it hasn't been initialized in the constructor

File: be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+            this.context.Groups.Add(group.FillCommonProperties());
+            var groupTranslation = new GroupTranslation
+            {
+                Name = model.Name,
+                Language = model.Language,
+                GroupId = group.Id
+            };
+
+            this.context.GroupTranslations.Add(groupTranslation.FillCommonProperties());
+            await this.context.SaveChangesAsync();
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs]: **suggestion:** Have you implemented the language fallback as in categories, etc.?

File: be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+                throw new CustomException(this.clientLocalizer.GetString("GroupNotFound"), (int)HttpStatusCode.NotFound);
+            }
+
+            var groupTranslation = await this.context.GroupTranslations.FirstOrDefaultAsync(x => x.GroupId == model.Id && x.Language == model.Language && x.IsActive);
+
+            if (groupTranslation is not null)
+            {
+                groupTranslation.Name = model.Name;
+                groupTranslation.LastModifiedDate = DateTime.UtcNow;
+            } else
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs]: **suggestion (non-blocking):** else should go to a new line. It's by convention in our C# code

File: be/src/Project/Services/Client/Client.Api/ServicesModels/Clients/CreateClientServiceModel.cs [hunk 16]
```diff
@@ -10,5 +11,6 @@ public class CreateClientServiceModel : BaseServiceModel
         public string CommunicationLanguage { get; set; }
         public string PhoneNumber { get; set; }
         public Guid? ClientOrganisationId { get; set; }
+        public IEnumerable<Guid> Groups { get; set; }
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/ServicesModels/Clients/CreateClientServiceModel.cs]: **suggestion:** Rename to ClientGroupIds

File: be/src/Project/Services/Client/Client.Api/Validators/Groups/GetGroupsModelValidator.cs [hunk 17]
```diff
@@ -0,0 +1,9 @@
+﻿using Client.Api.ServicesModels.Groups;
+using Foundation.Extensions.Validators;
+
+namespace Client.Api.Validators.Groups
+{
+    public class GetGroupsModelValidator : BasePagedServiceModelValidator<GetGroupsServiceModel>
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/Validators/Groups/GetGroupsModelValidator.cs]: **suggestion:** Rename to GetClientGroupsModelValidator like everywhere else

File: be/src/Project/Services/Client/Client.Api/v1/Controllers/GroupsController.cs [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+        [AllowAnonymous]
+        [ProducesResponseType((int)HttpStatusCode.OK, Type = typeof(PagedResults<IEnumerable<GroupResponseModel>>))]
+        [ProducesResponseType((int)HttpStatusCode.NotFound)]
+        [ProducesResponseType((int)HttpStatusCode.UnprocessableEntity)]
+        public async Task<IActionResult> Get(string ids, string searchTerm, int pageIndex, int itemsPerPage, string orderBy)
+        {
+            var sellerClaim = this.User.Claims.FirstOrDefault(x => x.Type == AccountConstants.Claims.OrganisationIdClaim);
+            var groupsIds = ids.ToEnumerableGuidIds();
+
+            if (groupsIds.Any())
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/v1/Controllers/GroupsController.cs]: **suggestion:** Rename to clientGroupIds like everywhere else

File: be/src/Project/Services/Client/Client.Api/v1/RequestModels/ClientRequestModel.cs [hunk 19]
```diff
@@ -10,5 +11,6 @@ public class ClientRequestModel : RequestModelBase
         public string CommunicationLanguage { get; set; }
         public string PhoneNumber { get; set; }
         public Guid? OrganisationId { get; set; }
+        public IEnumerable<Guid> Groups { get; set; }
```
Comment (dawiddworak88) [inline be/src/Project/Services/Client/Client.Api/v1/RequestModels/ClientRequestModel.cs]: **suggestion:** Rename to ClientGroupIds and everywhere where you use list of Guids

File: be/src/Project/Web/Seller/Seller.Web/Areas/Clients/Repositories/GroupsRepository.cs [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+            var warehousesRequestModel = new PagedRequestModelBase
+            {
+                PageIndex = PaginationConstants.DefaultPageIndex,
+                ItemsPerPage = PaginationConstants.DefaultPageSize
+            };
+
+            var apiRequest = new ApiRequest<PagedRequestModelBase>
+            {
+                Language = language,
+                Data = warehousesRequestModel,
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Clients/Repositories/GroupsRepository.cs]: **suggestion:** Copy-paste from warehouses. Look for it in this PR. It would be best to use simply 'requestModel', so that we don't have such issues by copy-paste anymore

File: fe/src/shared/components/MenuTiles/MenuTiles.scss [hunk 21]
```diff
@@ -1,10 +1,21 @@
 .menu-tiles {
-    display: flex;
-    flex-wrap: wrap;
-    justify-content: center;
+    margin: 0 auto;
```
Comment (dawiddworak88) [inline fe/src/shared/components/MenuTiles/MenuTiles.scss]: **suggestion:** let's remove these changes from this PR and do it in a separate one. I have a feeling we need to test it a lot.

File: fe/src/shared/components/MenuTiles/MenuTiles.js [hunk 22]
```diff
@@ -1,16 +1,57 @@
-import React from "react";
+import React, { useEffect } from "react";
 import PropTypes from "prop-types";
 import Tile from "./Tile";
 
 function MenuTiles(props) {
+    if (typeof window !== "undefined") {
```
Comment (dawiddworak88) [inline fe/src/shared/components/MenuTiles/MenuTiles.js]: **suggestion:** Let's move all of these changes to another PR.

File: fe/src/project/Seller.Portal/areas/Clients/pages/GroupPage/GroupPage.js [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+
+  toast.configure();
+
+  return (
+    <ThemeProvider theme={GlobalHelper.initMuiTheme()}>
+      <Store>
+        <div>
+          <Header {...props.header}></Header>
+          <MenuTiles {...props.menuTiles} />
+          <GroupForm {...props.groupForm} />
```
Comment (dawiddworak88) [inline fe/src/project/Seller.Portal/areas/Clients/pages/GroupPage/GroupPage.js]: suggestion: Rename to <ClientGroupForm {...props.clientGroupForm} /> like everywhere else

**Review by dawiddworak88 — CHANGES_REQUESTED** [review]

File: be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs (hunk 14 above)
Comment (SebaKas21) [reply be/src/Project/Services/Client/Client.Api/Services/Groups/GroupsService.cs]: @dawiddworak88  Yes :)

**Review by dawiddworak88 — APPROVED** [review]
Let's give it a try 👍
